### DISCUSSION TITLE
📝 [RFC] Unify allowClear api for clearable component

### DISCUSSION THREAD

**Discussion by kiner-tang** [ant-design/ant-design#43582] [RFCs - archive]
# 📝 [RFC] Unify allowClear api for clearable component

## Summary

Currently, most of our clearable components already support the `allowClear`  property to control whether a clear button is displayed to allow the user to manually clear the content/value of the component, but `allowClear`  is only a switch and users cannot customize the clear button.

The following components need to be refactored：

- [x] **AutoComplete** #43745
- [x] **Cascader** #43776
- [x] **DatePicker** https://github.com/react-component/picker/pull/660 #43811
- [x] **Input**
- [x] **Input.Textarea** #43836
- [x] **Select** https://github.com/react-component/select/pull/965
- [x] **TimePicker**  #43839
- [x] **TreeSelect** #43842

## Motivation

As #43495 mentioned.

 `allowClear` needs to be refactored in order to support customize clear button and facilitate subsequent API extensions.

## API

```typescript
type ComponentProps = {
  allowClear?: boolean | { clearIcon?: ReactNode }
}
```

## Basic Example

```tsx
<Input placeholder="input with clear icon" allowClear onChange={onChange} />
<Input placeholder="input with clear icon" allowClear={ clearIcon: <ClearOutlined /> } onChange={onChange} />
```

## Detail

react-componet needs to be refactored simultaneously. After react-component is refactored, antd upgrades corresponding react-component and updates related documents simultaneously.

**Comment by MadCcc** [comment]
`onClear` 不包含在 `allowClear` 中
